python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint_search.py

```diff
@@ -0,0 +1,1113 @@
+######################################################################
+# BioSimSpace: Making biomolecular simulation a breeze!
+#
+# Copyright: 2017-2022
+#
+# Authors: Lester Hedges <lester.hedges@gmail.com>
+#
+# BioSimSpace is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# BioSimSpace is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with BioSimSpace. If not, see <http://www.gnu.org/licenses/>.
+#####################################################################
+
+"""
+Functionality for selecting receptor-ligand restraints from unrestrained
+simulations.
+"""
+
+__author__ = "Finlay Clark"
+__email__ = "finlay.clark@ed.ac.uk"
+
+__all__ = ["RestraintSearch"]
+
+import sys as _sys
+import os as _os
+import tempfile as _tempfile
+import warnings as _warnings
+
+try:
+    import alchemlyb as _alchemlyb
+    from alchemlyb.postprocessors.units import R_kJmol, kJ2kcal
+    from alchemlyb.parsing.gmx import extract_u_nk as _gmx_extract_u_nk
+    from alchemlyb.parsing.gmx import extract_dHdl as _gmx_extract_dHdl
+    from alchemlyb.parsing.amber import extract_u_nk as _amber_extract_u_nk
+    from alchemlyb.parsing.amber import extract_dHdl as _amber_extract_dHdl
+    from alchemlyb.preprocessing.subsampling import statistical_inefficiency as _statistical_inefficiency
+    from alchemlyb.estimators import AutoMBAR as _AutoMBAR
+    from alchemlyb.estimators import TI as _TI
+    from alchemlyb.postprocessors.units import to_kcalmol as _to_kcalmol
+    is_alchemlyb = True
+except:
+    print('Please install alchemlyb via pip for analysis using it.')
+    is_alchemlyb = False
+
+try:
+    import MDRestraintsGenerator
+    from MDRestraintsGenerator import search as _search
+    from MDRestraintsGenerator.restraints import FindBoreschRestraint as _FindBoreschRestraint
+    is_MDRestraintsGenerator = True
+except:
+    print('Please install MDRestraintsGenerator for analysis using it.')
+    is_MDRestraintsGenerator = False
```

Comment:
Sorry, I'm struggling a bit on how to do this, currently, I did this with
```
_MDRestraintsGenerator = _try_import("MDRestraintsGenerator",
                                     install_command='pip install MDRestraintsGenerator')

from _MDRestraintsGenerator import search as _search
from _MDRestraintsGenerator.restraints import FindBoreschRestraint as _FindBoreschRestraint
```
This will fail with
```
Traceback (most recent call last):
  File "/Users/zwu/miniconda3_x86/envs/exs-orionbssfloes/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3433, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-90e5ed732635>", line 1, in <module>
    from _MDRestraintsGenerator import search as _search
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydev_bundle/pydev_import_hook.py", line 21, in do_import
    module = self._system_import(name, *args, **kwargs)
ModuleNotFoundError: No module named '_MDRestraintsGenerator'
```
The _MDRestraintsGenerator seems to be the right module though
```
_MDRestraintsGenerator
Out[4]: <module 'MDRestraintsGenerator' from '/Users/zwu/miniconda3_x86/envs/exs-orionbssfloes/lib/python3.8/site-packages/MDRestraintsGenerator/__init__.py'>
```
`from MDRestraintsGenerator import search as _search` would work fine.

Reply:
You can't import and rename, then use the renamed module to import from. It's just a local variable name, so isn't used by the import system.

I think something like the following is what we want:

```
_MDRestraintsGenerator = _try_import("MDRestraintsGenerator",
                                     install_command='pip install MDRestraintsGenerator')
                                     

if _have_imported(_MDRestraintsGenerator):
    from MDRestraintsGenerator import search as _search
    from MDRestraintsGenerator.restraints import FindBoreschRestraint as _FindBoreschRestraint
```

This way, we only do the `from X import Y` bit if the `_try_import("X")` bit worked.